oeps/oep-0052-arch-event-bus-technology.rst

```diff
@@ -0,0 +1,264 @@
+
+.. _Amazon MSK: https://aws.amazon.com/msk/
+.. _Confluent Platform: https://www.confluent.io/product/confluent-platform
+.. _comparison of Apache Kafka vs Vendors: https://www.kai-waehner.de/blog/2021/04/20/comparison-open-source-apache-kafka-vs-confluent-cloudera-red-hat-amazon-msk-cloud/
+
+Kafka Highlights
+~~~~~~~~~~~~~~~~
+
+Pros
+^^^^
+
+* Battle-tested, widely adopted, big community, lots of documentation and answers.
+* Amazon MSK (AWS service) provides hosted path of least resistance.
+* `New Relic integration with Amazon MSK`_ (useful to edX.org).
+
+Cons
+^^^^
+
+* Many open questions about add-ons required for developers and operators.
+* Complex to manage, including likely manual scaling.
+
+.. _New Relic integration with Amazon MSK: https://docs.newrelic.com/docs/integrations/amazon-integrations/aws-integrations-list/aws-managed-kafka-msk-integration/
+
+Consequences
+------------
+
+* Operators will need to deploy and manage the selected infrastructure, which is likely to be complex. If Apache Kafka is selected, there are likely to be a set of auxiliary parts to provide all required functionality for our message bus.
+* Education will be required for both developers and operators regarding best practices for each role.
+* Code to interact with Kafka and its libraries will be added to core services.
+* At least one initial use case must be completed. One potential candidate is the grade change event in the LMS, and its use by the Credentials service.
+* Once we have a message bus, we can investigate other potential use cases:
+
+  * Course/program update propagation.
+  * Feed into xAPI/Caliper capabilities.
+  * New services and features can be built fully de-coupled from the core application.
+
+Rejected Alternatives
+---------------------
+
+Apache Pulsar
+~~~~~~~~~~~~~
+
+Although rejected to start, `Apache Pulsar`_ remains an option if solving with Kafka turns out to be overly burdensome for developers or operators.
+
+Pros
+^^^^
+
+* Ease of scalability (built-in, according to docs).
+* Ease of data retention capabilities.
+* Additional built-in pub/sub features (built-in, according to docs).
+
+Cons
+^^^^
+
+* Requires 3rd party hosting or larger upfront investment in self-hosted (kubernetes).
+* Less mature (but growing) community, little documentation, and few answers.
+
+Note: Read an interesting (Kafka/Confluent) biased article exploring `comparisons and myths of Kafka vs Pulsar`_.
+
+.. _Apache Pulsar: https://pulsar.apache.org/
+.. _comparisons and myths of Kafka vs Pulsar: https://dzone.com/articles/pulsar-vs-kafka-comparison-and-myths-explored
+
+Redis
+~~~~~
+
+Pros
+^^^^
+
+* Already part of Open edX platform
+
+Cons
+^^^^
+
+* Can lose acked data, even if RAM backed up with an append-only file (AOF).
+* Requires homegrown schema management.
+
+Abstract Message Bus Class
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+@feanil - What is this???
```

Comment:
This is the idea of building an abstraction for the messaging tool we use so that the edx-platform can remain agnostic and just define an interface that can be implemented with either Kafka or Pulsar.  This allows people to have flexibility in choosing their backends.  The cost here is that we would need to build and maintain the abstraction and it would prevent us from using unique features available with whatever backend we chose.

Reply:
Right, but a counterpoint to haystack is celery/kombu, which gave us an easy path from RabbitMQ to Redis, which has been operationally much simpler. Another is our own internal abstraction of `MongoConnection`, which is enabling recent work to remove the MongoDB dependency from the split modulestore. I think it mostly boils down to how large the interface we have to support actually is, and (as @feanil said) what unique features we lose. It seems like Pulsar and Redis Streams both drew heavy inspiration from Kafka, so I'm optimistic that it's a manageable layer. We'd probably want something like that just to be able to swap it out for a mockable interface when running tests anyhow.

Reply:
Agreed. Not much more to write on this topic at this time. We need to make some headway on some other fronts, and then we'll have a little more idea about what would need to be abstracted and how and if that could work. However, in addition to earlier mentioned possible downsides, this would also take additional resources to prove out that the abstraction really could work with Redis Streams. Maybe we could line up wider community resources for that type of effort?

Reply:
I like where this conversation is going. To summarize the arguments:

1. We need to define what features will be needed from our event bus to figure out whether an abstraction layer can be built.
2. The selected event bus will have to fit in a devstack.
3. We need a good deployment story for most Open edX hosting platforms, and not just edX.org.

Reply:
An update from myself and @robrap as we work more with this, we are starting to see that the API may not be as complex as we initially thought and an abstraction layer may be more viable.  No promises yet but , it's something we'll be looking at more as we build a working version of this without the abstraction layer.

@regisb thanks for the summary, I think it captures the current conversation very well.  One thought I had is reasonably quantifying "good" in the 3rd point.  I think this is perhaps a good conversation independent of this particular technology conversation so I've re-framed/[asked that question in the forums](https://discuss.openedx.org/t/a-good-deployment-story-for-most-open-edx-hosting-platforms/5915).